File: ios/Wispers Access/Demo/DemoMode.swift
import Foundation
import WispersAccessSdk

/// Static roster for App Store screenshots: replaces the SDK's store and the
/// live host polling with fixed shares and statuses, so captures are
/// deterministic and need no backend. Debug-only — the launch argument is
/// compiled out of release builds.
///
/// Activate with the `--demo` launch argument (an Xcode scheme, or
/// `xcrun simctl launch booted <bundle-id> --demo`).
@MainActor
enum DemoMode {
    static let active: Bool = {
        #if DEBUG
            return ProcessInfo.processInfo.arguments.contains("--demo")
        #else
            return false
        #endif
    }()

    /// What the demo manager throws for anything that needs a host.
    struct NotAvailable: LocalizedError {
        var errorDescription: String? { "Not available in demo mode." }
    }

    /// A ShareManager without an SDK client: the fixed roster, and icons and
    /// activity in a throwaway temp directory — the real files are never
    /// touched.
    static func makeManager() -> ShareManager {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("demo-roster")
        // A previous run's seed would resurface with stale timestamps.
        try? FileManager.default.removeItem(at: dir)
        let icons = ShareIconStore(fileURL: dir.appendingPathComponent("share-icons.json"))
        let activity = ShareActivityStore(fileURL: dir.appendingPathComponent("share-activity.json"))
        let now = Date()
        var shares: [Share] = []
        for entry in roster {
            shares.append(
                Share(
                    id: entry.slug,
                    name: entry.nickname,
                    label: entry.slug,
                    transport: .iroh,
                    apps: [SharedApp(id: "app", name: entry.nickname, kind: .web)],
                    state: .live,
                    joinedAt: now.addingTimeInterval(-entry.joined)
                )
            )
            activity.markConnected(entry.slug, at: now.addingTimeInterval(-entry.lastConnected))
            if let url = Bundle.main.url(forResource: entry.slug, withExtension: "png"),
                let png = try? Data(contentsOf: url)
            {
                icons.update(png, rank: 1, for: BrowseKey(shareID: entry.slug, appID: "app"))
            }
        }
        return ShareManager(client: nil, icons: icons, activity: activity, shares: shares)
    }

    /// A screen to open on launch (`--demo-detail <slug>`, e.g. `grafana`), so
    /// capture scripts can shoot the detail screen without synthesizing taps.
    static var initialRoute: ShareRoute? {
        guard active else { return nil }
        let args = ProcessInfo.processInfo.arguments
        guard let flag = args.firstIndex(of: "--demo-detail"), flag + 1 < args.count else {
            return nil
        }
        return .detail(args[flag + 1])
    }

    /// Whether to open the add-app sheet on launch (`--demo-add`), pre-filled
    /// with `sampleInvite`, so capture scripts can shoot the code-entry screen.
    static var presentAddSheet: Bool {
        active && ProcessInfo.processInfo.arguments.contains("--demo-add")
    }

    /// Shaped like a real `waserver invite` code for an iroh share (endpoint
    /// id, secret) but pure fiction — it parses, so the screenshot shows the
    /// transport note, and joins nothing.
    static let sampleInvite =
        "wax1_iroh_30b1c2fed7e381856aad2334030bd0cf4c316ad4ee2fb08a4eebe661718c5977_d4ab8faf8841f2790da9bc82ee370ec7"

    /// Fixed per-share availability, replacing the host poll.
    static var statuses: [ShareId: Availability] {
        Dictionary(uniqueKeysWithValues: roster.map { ($0.slug, $0.availability) })
    }

    // Known self-hosted tools anchor the "that's my stack" reaction; one bespoke
    // entry shows shares aren't limited to famous products. Names are nominative
    // word-mark use; the icons are our own brand-colored glyphs, not the logos.
    private static let roster = [
        Entry(
            nickname: "Stats (Grafana)",
            slug: "grafana",
            availability: .online,
            lastConnected: 2 * 60,
            joined: 24 * day
        ),
        Entry(
            nickname: "ERP (Odoo)",
            slug: "odoo",
            availability: .online,
            lastConnected: 60 * 60,
            joined: 18 * day
        ),
        Entry(
            nickname: "Files (Nextcloud)",
            slug: "nextcloud",
            availability: .online,
            lastConnected: 30 * 60,
            joined: 11 * day
        ),
        Entry(
            nickname: "Duty Roster",
            slug: "duty-roster",
            availability: .offline,
            lastConnected: 1 * day,
            joined: 5 * day
        ),
    ]

    private static let day: TimeInterval = 24 * 60 * 60

    private struct Entry {
        let nickname: String
        let slug: String
        let availability: Availability
        let lastConnected: TimeInterval
        let joined: TimeInterval
    }
}

File: android/app/src/main/java/dev/wispers/access/android/demo/DemoMode.kt
package dev.wispers.access.android.demo

import android.app.Activity
import android.content.pm.ApplicationInfo
import dev.wispers.access.android.proxy.ShareAvailability
import dev.wispers.access.android.storage.Share
import dev.wispers.access.android.storage.ShareId
import java.time.Duration
import java.time.Instant

/**
 * Static roster for store screenshots: replaces the Room-backed roster and the
 * live hub polling with fixed shares and statuses, so captures are
 * deterministic and need no backend. Debug-only — the launch extra is ignored
 * in release builds, and the icon assets ship only in the debug source set.
 *
 * Activate with:
 *
 *     adb shell am start -n dev.wispers.access.android/.MainActivity --ez demo true
 */
object DemoMode {
    /** The fixed roster, or null when demo mode is off (the normal case). */
    var shares: List<Share>? = null
        private set

    /** Fixed per-share availability, replacing the hub poll. */
    var statuses: Map<ShareId, ShareAvailability> = emptyMap()
        private set

    val active: Boolean get() = shares != null

    fun maybeActivate(activity: Activity) {
        if (!activity.intent.getBooleanExtra(EXTRA_DEMO, false)) return
        if (activity.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE == 0) return
        val now = Instant.now()
        shares = ROSTER.map { it.toShare(activity, now) }
        statuses = ROSTER.associate { ShareId(it.slug) to it.availability }
    }

    private const val EXTRA_DEMO = "demo"

    // Known self-hosted tools anchor the "that's my stack" reaction; one bespoke
    // entry shows shares aren't limited to famous products. Names are nominative
    // word-mark use; the icons are our own brand-colored glyphs, not the logos.
    private val ROSTER = listOf(
        DemoShare(
            nickname = "Stats (Grafana)",
            slug = "grafana",
            availability = ShareAvailability.ONLINE,
            lastConnected = Duration.ofMinutes(2),
            joined = Duration.ofDays(24),
        ),
        DemoShare(
            nickname = "ERP (Odoo)",
            slug = "odoo",
            availability = ShareAvailability.ONLINE,
            lastConnected = Duration.ofHours(1),
            joined = Duration.ofDays(18),
        ),
        DemoShare(
            nickname = "Files (Nextcloud)",
            slug = "nextcloud",
            availability = ShareAvailability.ONLINE,
            lastConnected = Duration.ofMinutes(30),
            joined = Duration.ofDays(11),
        ),
        DemoShare(
            nickname = "Duty Roster",
            slug = "duty-roster",
            availability = ShareAvailability.OFFLINE,
            lastConnected = Duration.ofDays(1),
            joined = Duration.ofDays(5),
        ),
    )

    private data class DemoShare(
        val nickname: String,
        val slug: String,
        val availability: ShareAvailability,
        val lastConnected: Duration,
        val joined: Duration,
    ) {
        fun toShare(activity: Activity, now: Instant) = Share(
            id = ShareId(slug),
            nickname = nickname,
            createdAt = now - joined,
            lastConnectedAt = now - lastConnected,
            iconPng = activity.assets.open("demo/$slug.png").readBytes(),
            terminalState = null,
        )
    }
}
